// @dependency/javascriptTestRunner/source/runInContainer/script.js
"use strict";var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.unitTest = unitTest;var _path = _interopRequireDefault(require("path"));
const { execSync, spawn, spawnSync } = require('child_process');
const deploymentScriptPath = _path.default.dirname(require.resolve('@dependency/deploymentScript/package.json')),
javascriptTestRunnerPath = _path.default.normalize(`${__dirname}/../entrypoint/cli/transpilation.entrypoint.js`);







function unitTest(
input)


{
  let configuration,
  container = {},
  nodeFlag = {},
  testRunnerModulePath,
  testPath,
  applicationPathOnHostMachine;
  ({
    configuration,
    container = {
      imageName: container.imageName = configuration.dockerImageName,
      ymlFile: container.ymlFile = `${deploymentScriptPath}/deploymentContainer/deployment.dockerCompose.yml` },

    nodeFlag = {
      debug: nodeFlag.debug = null,
      break: nodeFlag.break = null },

    testRunnerModulePath = javascriptTestRunnerPath,
    testPath = configuration.directory.testPath,
    applicationPathOnHostMachine = _path.default.join(configuration.directory.projectPath, 'application') } =
  input);

  let serviceName = 'nodejs',
  containerPrefix = 'app_test';

  console.log(`\x1b[33m\x1b[1m\x1b[7m\x1b[36m%s\x1b[0m \x1b[2m\x1b[3m%s\x1b[0m`, `Running Container:`, `NodeJS App`);
  let debugCommand = nodeFlag.debug ? `--inspect${nodeFlag ? '-brk' : ''}=0.0.0.0:9229` : '';
  let appEntrypointPath = testRunnerModulePath;
  let firstNodeCommand = testPath;


  let containerCommand = `node ${debugCommand} ${appEntrypointPath} ${firstNodeCommand}`;

  let environmentVariable = {
    DEPLOYMENT: 'development',
    SZN_DEBUG: false,
    applicationPathOnHostMachine,
    imageName: container.imageName };

  let processCommand = 'docker-compose',
  processCommandArgs = [`-f ${container.ymlFile}`, `--project-name ${containerPrefix}`, `run --service-ports --use-aliases`, `--entrypoint "${containerCommand}"`, `${serviceName}`],
  processOption = {

    shell: true,
    detached: false,
    stdio: ['inherit', 'inherit', 'inherit', 'ipc'],
    env: Object.assign(
    process.env,
    environmentVariable) };


  console.log(`%s %s`, processCommand, processCommandArgs.join(' '));
  let childProcess = spawn(processCommand, processCommandArgs, processOption);
  childProcess.on('error', function (err) {
    throw err;
  });
  childProcess.on('exit', () => {
    console.log(`PID: Child ${childProcess.pid} terminated.`);
  });

  console.log(`PID: Child ${childProcess.pid}`);
  process.on('SIGINT', () => {

    console.log('• Caught interrupt signal - host machine level');
    childProcess.kill('SIGINT');
  });
}
