// @dependency/javascriptTranspilation/source/utility/transformPlugin.babel.js
"use strict";var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");Object.defineProperty(exports, "__esModule", { value: true });exports.transformNamedModuleToPath = transformNamedModuleToPath;exports.minifyHtmlTemplateLiterals = minifyHtmlTemplateLiterals;var babel = _interopRequireWildcard(require("@babel/core"));







function transformNamedModuleToPath() {
  return {
    visitor: {
      'ImportDeclaration|ExportNamedDeclaration'(path) {
        let source = path.node.source;
        if (!source) return;

        let newSourceValue;
        if (source.value.startsWith('@') || !source.value.startsWith('/') && !source.value.startsWith('.')) {

          newSourceValue = `/@webcomponent/@package/${source.value}`;
          source.value = newSourceValue;
        }
      } } };


}




function minifyHtmlTemplateLiterals() {

  return {
    visitor: {
      TaggedTemplateExpression(path) {
        let tagName = path.node.tag.name;
        if (!(tagName == 'html')) return;
        path.node.quasi.expressions;
        let templateLiteral = path.node.quasi;
        let expression = templateLiteral.expressions;
        if (expression) {
          for (let index in expression) {
            let code = babel.transformFromAst({
              type: 'File',
              program: {
                type: 'Program',
                sourceType: 'module',
                body: [expression[index]] } });


            console.log(code);
          }
        }
      } } };


}
