// babel-plugin-parameter-decorator/test/lib/ts/Greeter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _decorators = require("../decorators");

var _Sentinel = _interopRequireDefault(require("./Sentinel"));

var _class, _class2, _temp;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) { var desc = {}; Object['ke' + 'ys'](descriptor).forEach(function (key) { desc[key] = descriptor[key]; }); desc.enumerable = !!desc.enumerable; desc.configurable = !!desc.configurable; if ('value' in desc || desc.initializer) { desc.writable = true; } desc = decorators.slice().reverse().reduce(function (desc, decorator) { return decorator(target, property, desc) || desc; }, desc); if (context && desc.initializer !== void 0) { desc.value = desc.initializer ? desc.initializer.call(context) : void 0; desc.initializer = undefined; } if (desc.initializer === void 0) { Object['define' + 'Property'](target, property, desc); desc = null; } return desc; }

var Greeter = (0, _decorators.Factory)(_class = (_class2 = (_temp =
/*#__PURE__*/
function () {
  function Greeter(greeting, sentinel) {
    _classCallCheck(this, Greeter);

    this.greeting = greeting;
    this.sentinel = sentinel;
    this.counter = this.sentinel.counter;
  }

  Greeter = (0, _decorators.Inject)(_Sentinel.default)(Greeter, undefined, 1);

  _createClass(Greeter, [{
    key: "greet",
    value: function greet(name, emoj) {
      this.sentinel.count();
      var greeting = 'how are you?';
      return "Hello " + name + ", " + (this.greeting || greeting);
    }
  }, {
    key: "welcome",
    value: function welcome(firstName, lastName) {
      this.sentinel.count();
      return "Welcome " + lastName + "." + firstName;
    }
  }, {
    key: "count",
    value: function count() {
      return this.counter.number;
    }
  }]);

  (0, _decorators.required)('lastName')(Greeter.prototype, "welcome", 1);
  (0, _decorators.required)('firstName')(Greeter.prototype, "welcome", 0);
  (0, _decorators.optional)(Greeter.prototype, "greet", 1);
  (0, _decorators.required)('name')(Greeter.prototype, "greet", 0);
  return Greeter;
}(), _temp), (_applyDecoratedDescriptor(_class2.prototype, "greet", [_decorators.validate], Object.getOwnPropertyDescriptor(_class2.prototype, "greet"), _class2.prototype), _applyDecoratedDescriptor(_class2.prototype, "welcome", [_decorators.validate], Object.getOwnPropertyDescriptor(_class2.prototype, "welcome"), _class2.prototype)), _class2)) || _class;

var _default = Greeter;
exports.default = _default;
