// @dependency/javascriptTranspilation/source/additionalRequireHook.js
"use strict";const { transformFileSync } = require('@babel/core'),
{ addHook: addRequireHook } = require('pirates'),
path = require('path'),
filesystem = require('fs'),
assert = require('assert'),
{ removeMatchingStringFromBeginning } = require('./utility/removeMatchingStringFromBeginning.js'),
defaultOutputRelativePath = './temporary/transpiled',
isPreservedSymlink = require('./utility/isPreservedSymlinkFlag.js');

function filesystemTranspiledOutput({
  babelConfig,
  shouldTransform = false,
  targetProjectConfig,
  outputRelativePath = defaultOutputRelativePath,
  extension,
  ignoreNodeModules = false,
  ignoreFilenamePattern = [] })
{

  outputRelativePath = targetProjectConfig.transpilation && targetProjectConfig.transpilation.outputDirectory || outputRelativePath;
  addRequireHook(
  (code, filename) => {
    let content;

    if (shouldTransform) {
    } else {
      content = code;
    }


    let outputPath = path.join(targetProjectConfig.rootPath, outputRelativePath);
    let relativeFilePath = removeMatchingStringFromBeginning({ basePath: targetProjectConfig.rootPath, targetPath: filename });
    let traspiledFilePath = path.join(outputPath, relativeFilePath);

    filesystem.mkdirSync(path.dirname(traspiledFilePath), { recursive: true });

    filesystem.writeFileSync(traspiledFilePath, content, { encoding: 'utf8' });
    return code;
  },
  {
    exts: extension,
    ignoreNodeModules: ignoreNodeModules,
    matcher: filename => ignoreFilenamePattern.some(regex => filename.match(regex)) ? false : true });


}

function transformHook({ extension, ignoreNodeModules = false, ignoreFilenamePattern = [] } = {}) {
  addRequireHook(
  (code, filename) => {
    let content;

    let transformed = transformFileSync(
    filename,
    Object.assign(
    {

      sourceMaps: 'both' || 'inline' || true,
      ast: false },

    babelConfig));



    content = transformed.code;
    return content;
  },
  {
    exts: extension,
    ignoreNodeModules: ignoreNodeModules,
    matcher: filename => ignoreFilenamePattern.some(regex => filename.match(regex)) ? false : true });


}

module.exports = { transformHook, filesystemTranspiledOutput };
