// @dependency/entity/source/Constructable/Constructable.class.js
"use strict";var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.Constructable = void 0;var _skipFirstGeneratorNext2 = _interopRequireDefault(require("@babel/runtime/helpers/skipFirstGeneratorNext"));var _SymbolReference = require("../functionalityPrototype/Symbol.reference.js");
var _exportFunctionality = require("../functionalityPrototype/exportFunctionality.js");
var _createObjectWithDelegation = require("../utility/createObjectWithDelegation.js");
var _generatorExecutionControl = require("../utility/generatorExecutionControl.js");
var _mergeProperty = require("../utility/mergeProperty.js");
var lodash = _interopRequireWildcard(require("lodash"));


const Reference = Object.assign(
Object.create(Object.prototype),
{
  name: Symbol('Own class name'),
  class: Symbol('class'),
  reference: 'reference',
  metadata: _SymbolReference.metadata,
  key: {
    createObjectWithDelegation: Symbol('createObjectWithDelegation'),
    classInstance: Symbol('class instance related'),
    configuredClass: Symbol('Configured class/constructable related'),
    constructableClass: Symbol('Constructable related') } },


_exportFunctionality.instantiateInitialize.Reference,
_exportFunctionality.constructor.Reference,
_exportFunctionality.clientInterface.Reference);


const Prototype = Object.assign(Object.create(Object.prototype), _exportFunctionality.instantiateInitialize.Prototype, _exportFunctionality.constructor.Prototype, _exportFunctionality.clientInterface.Prototype);










Prototype[Reference.prototypeDelegation.functionality].setter.call(Prototype, {
  [Reference.key.constructableClass]: {
    prototype: Prototype,
    reference: Reference } });











Prototype[Reference.instantiate.functionality].setter.call(Prototype, {

  [Reference.key.createObjectWithDelegation]: _createObjectWithDelegation.createObjectWithDelegation });








Prototype[Reference.initialize.functionality].setter.call(Prototype, {

  [Reference.key.constructableClass]: function initializeConstuctable({ targetInstance, reference, prototype, description, construtorProperty = this } = {}) {
    if (!reference) {
      let prototypeDelegationSetting = construtorProperty[Reference.prototypeDelegation.functionality].getter.call(construtorProperty, Reference.key.constructableClass);
      reference = Object.create(prototypeDelegationSetting.reference || null);
    }
    if (!prototype) {
      let prototypeDelegationSetting = construtorProperty[Reference.prototypeDelegation.functionality].getter.call(construtorProperty, Reference.key.constructableClass);
      prototype = Object.create(prototypeDelegationSetting.prototype || null);
    }
    Object.setPrototypeOf(targetInstance, prototype);

    targetInstance[Reference.prototypeDelegation.functionality].setter.call(targetInstance, {
      [Reference.key.constructableClass]: {
        prototype: prototype,
        reference: reference } });


    let prototypeDelegationGetter = targetInstance[Reference.prototypeDelegation.functionality].getter.bind(targetInstance);
    (0, _mergeProperty.mergeNonexistentProperties)(targetInstance, {





      get reference() {
        return prototypeDelegationGetter(Reference.key.constructableClass).reference;
      },



      [Reference.class]: construtorProperty });

    Object.defineProperty(targetInstance, Reference.name, { writable: false, enumerable: false, value: description });

    if (!prototype.hasOwnProperty(Reference.metadata))
    Object.defineProperty(prototype, Reference.metadata, { writable: false, enumerable: false, value: { type: Symbol(`${description} functionality`) } });
    return targetInstance;
  },

  [Reference.key.classInstance]({ targetInstance, construtorProperty = this, description = '' } = {}) {
    if (!targetInstance.hasOwnProperty(Reference.metadata))
    Object.defineProperty(targetInstance, Reference.metadata, {
      writable: false,
      enumerable: false,
      value: { type: Symbol(`${description} - created by ${construtorProperty[Reference.name]} class`) } });

    targetInstance.constructor = construtorProperty;
    return targetInstance;
  },
  [Reference.key.configuredClass]({ targetInstance, parameter = [] } = {}) {
    targetInstance.parameter = parameter;
    return targetInstance;
  } });









Prototype[Reference.constructor.functionality].setter.call(Prototype, {



  [Reference.key.constructableClass]: (function () {let _ref = function* ({ description, reference, prototype, prototypeDelegation, callerClass = this } = {}) {let _functionSent = yield;
      const shouldHandOverControl = _generatorExecutionControl.executionControl.shouldHandOver(_functionSent);
      prototypeDelegation || (prototypeDelegation = callerClass[Reference.prototypeDelegation.functionality].getter.call(callerClass, Reference.key.constructableClass).prototype);
      const step = [

      {
        passThroughArg: { description, prototypeDelegation: prototypeDelegation },
        func: function (previousArg, arg) {var _context;
          let instance = (_context = callerClass, callerClass[Reference.instantiate.functionality].switch).call(_context, { implementationKey: Reference.key.createObjectWithDelegation })(arg);
          return { instance };
        },
        condition: true },

      {
        passThroughArg: { description, reference, prototype, construtorProperty: callerClass },
        func: function ({ instance }, arg) {var _context2;

          let argumentObject = Object.assign({ targetInstance: instance }, arg);
          (_context2 = callerClass, callerClass[Reference.initialize.functionality].switch).call(_context2, { implementationKey: Reference.key.constructableClass, recursiveDelegationChainExecution: true })(argumentObject);
          (_context2 = callerClass, callerClass[Reference.initialize.functionality].switch).call(_context2, { implementationKey: Reference.key.classInstance })(
          lodash.pick(argumentObject, ['description', 'targetInstance', 'construtorProperty']));

          return instance;
        },
        condition: true }];




      let i = 0,
      result;
      while (i < step.length) {
        if (step[i].condition == false) {
          i++;
          continue;
        }
        if (shouldHandOverControl) {
          _functionSent = yield step[i].passThroughArg;
          result = step[i].func(result, _functionSent);
        } else {
          result = step[i].func(result, step[i].passThroughArg);
        }
        i++;
      }
      return result;
    },_ref2 = (0, _skipFirstGeneratorNext2.default)(_ref);return new Proxy(_ref, { apply(target, thisArgument, argumentsList) {return Reflect.apply(_ref2, thisArgument, argumentsList);} });})(),
  [Reference.key.classInstance]({ description = 'Class', callerClass = this } = {}) {
    let instance = callerClass[Reference.instantiate.functionality].switch.call(callerClass, { implementationKey: Reference.key.createObjectWithDelegation })({
      description,
      prototypeDelegation: callerClass });

    callerClass[Reference.initialize.functionality].switch.call(callerClass, { implementationKey: Reference.key.classInstance })({ targetInstance: instance, description: description });
    return instance;
  },

  [Reference.key.configuredClass]({ description = 'Configured Class.', parameter, callerClass = this } = {}) {
    let instance = callerClass[Reference.constructor.functionality].switch.call(callerClass, { implementationKey: Reference.key.classInstance })({ description: description });
    callerClass[Reference.initialize.functionality].switch.call(callerClass, { implementationKey: Reference.key.configuredClass })({ targetInstance: instance, parameter });
    return instance;
  } });












const Constructable = Prototype[Reference.constructor.functionality].switch.call(Prototype, { implementationKey: Reference.key.constructableClass })({
  description: 'Constructable',
  reference: Reference,
  prototype: Prototype });exports.Constructable = Constructable;









let prototypeDelegationGetter = Constructable[Reference.prototypeDelegation.functionality].getter;
Prototype[Reference.clientInterface.functionality].setter.call(Prototype, {





  [Reference.key.constructableClass]({
    constructorImplementation,
    returnedInstanceAdapter = instance => ({
      class: instance,
      reference: instance[Reference.reference],
      constructablePrototype: prototypeDelegationGetter.call(instance, Reference.key.constructableClass).prototype }),

    callerClass = this } =
  {}) {
    let constructorSwitch = Constructable[Reference.constructor.functionality].switch,
    clientInterfaceSwitch = Constructable[Reference.clientInterface.functionality].switch,
    prototypeDelegationGetter = Constructable[Reference.prototypeDelegation.functionality].getter;
    let configuredConstructableImplementation = Reference.key.configuredClass,
    clientInterfaceImplementation = Reference.key.constructableClass;

    const proxiedTarget = new Proxy(function () {}, {
      construct(target, argumentList, proxiedTarget) {
        if (callerClass.parameter) (0, _mergeProperty.mergeArrayWithObjectItem)({ listTarget: argumentList, listDefault: callerClass.parameter });
        let instance = constructorSwitch.call(callerClass, { implementationKey: constructorImplementation })(...argumentList);
        return returnedInstanceAdapter ? returnedInstanceAdapter(instance) : instance;
      },
      apply(target, thisArg, [{ description, parameter = [] } = {}]) {
        let newConfiguredConstructable = constructorSwitch.call(callerClass, { implementationKey: configuredConstructableImplementation })({ description: description, parameter });
        let clientInterface = clientInterfaceSwitch.call(newConfiguredConstructable, { implementationKey: clientInterfaceImplementation })();
        return clientInterface;
      } });

    return proxiedTarget;
  } });


Constructable.clientInterface = Constructable[Reference.clientInterface.functionality].switch.call(Constructable, { implementationKey: Reference.key.constructableClass })({
  constructorImplementation: Reference.key.constructableClass });
