// @dependency/javascriptTranspilation/source/compilerConfiguration/serverRuntime.BabelConfig.js
"use strict";const path = require('path');

module.exports = {
  babelConfig: {

    sourceMaps: 'both' || 'inline' || true,
    retainLines: true,
    minified: false,
    comments: false,

    presets: [],

    plugins: [

    require.resolve('@babel/plugin-syntax-dynamic-import'),
    require.resolve('@babel/plugin-syntax-typescript'),






    require.resolve(`@babel/plugin-transform-runtime`),



    require.resolve('@babel/plugin-proposal-export-namespace-from'),
    require.resolve('@babel/plugin-proposal-export-default-from'),

    [require.resolve('@babel/plugin-transform-typescript'), {}],
    require.resolve(`@babel/plugin-transform-modules-commonjs`),
    require.resolve(`babel-plugin-dynamic-import-node`),
    [require.resolve(`@babel/plugin-proposal-decorators`), { legacy: true }],
    [require.resolve(`@babel/plugin-proposal-class-properties`), { loose: true }],
    require.resolve(`babel-plugin-parameter-decorator`),
    require.resolve(`@babel/plugin-proposal-function-bind`),
    require.resolve(`@babel/plugin-proposal-optional-chaining`),
    require.resolve(`@babel/plugin-proposal-object-rest-spread`),
    [require.resolve('@babel/plugin-proposal-logical-assignment-operators')],
    require.resolve(`@babel/babel-plugin-functionSentProxyImplementation`),

    [require.resolve('@babel/plugin-proposal-throw-expressions')],
    require.resolve('@babel/plugin-proposal-nullish-coalescing-operator'),
    [require.resolve('@babel/plugin-proposal-pipeline-operator'), { proposal: 'minimal' }],
    require.resolve('@babel/plugin-proposal-optional-catch-binding')] } };
