// @dependency/commonPattern/source/decoratorUtility.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.add = add;exports.execute = execute;exports.conditional = conditional;exports.executeOnceForEachInstance = executeOnceForEachInstance;exports.executionLevel = executionLevel;function add({ to = 'static' }, method) {
  return Class => {
    let targetReference;
    switch (to) {
      case 'prototype':
        targetReference = Class.prototype;
        break;
      case 'static':
      default:
        targetReference = Class;
        break;}

    Object.entries(method).forEach(
    ([key, value]) => targetReference[key] = value);

    return Class;
  };
}

function execute({ staticMethod, self = true, args = [] }) {

  return targetClass => {
    if (self) args.unshift(targetClass);
    targetClass[staticMethod](...args);
    return targetClass;
  };
}


function conditional({ condition = true, decorator }) {
  return condition ? decorator : Class => {return Class;};
}





function executeOnceForEachInstance() {
  return (target, methodName, descriptor) => {
    let method = target[methodName];
    descriptor.value = new Proxy(method, {
      apply: async (target, thisArg, argumentsList) => {
        if (thisArg.executedmethod &&
        thisArg.executedmethod[methodName] &&
        thisArg.executedmethod[methodName]['executed'])
        {
          return thisArg.executedmethod[methodName]['result'];
        }
        thisArg.executedmethod = {};
        thisArg.executedmethod[methodName] = {};
        let instance = await target.apply(thisArg, argumentsList);
        thisArg.executedmethod[methodName]['executed'] = true;
        thisArg.executedmethod[methodName]['result'] = instance;
        return instance;
      } });

    return descriptor;
  };
}





function executionLevel() {
  return (target, methodName, descriptor) => {
    let method = target[methodName];
    descriptor.value = new Proxy(method, {
      apply: async (target, thisArg, argumentsList) => {
        thisArg.executionLevel = thisArg.key ? 'nested' : 'topLevel';
        return await target.apply(thisArg, argumentsList);
      } });

    return descriptor;
  };
}
