// @dependency/javascriptTestRunner/source/runInContainer/clientInterface/commandLine.deprecated.js
"use strict";var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.cliAdapter = cliAdapter;


var _path = _interopRequireDefault(require("path"));
var _parseKeyValuePairSeparatedBySymbol = require("@dependency/parseKeyValuePairSeparatedBySymbol");
var _script = require("../script.js");


if (require.main === module) {
  console.log('• Executed directly.');
  cliAdapter();
}







function cliAdapter({ testType, configurationPath } = {}) {
  console.group('• Running entrypoint application in Manager Container:');
  console.log(`- passed process arguments: ${JSON.stringify(process.argv)}`);
  const namedArgs = (0, _parseKeyValuePairSeparatedBySymbol.parseKeyValuePairSeparatedBySymbolFromArray)({ array: process.argv });

  testType = testType || namedArgs['testType'] || null;


  if (!configurationPath) configurationPath = namedArgs.configuration ? _path.default.join(process.cwd(), namedArgs.configuration) : _path.default.join(process.cwd(), 'configuration');
  const configuration = require(configurationPath);

  switch (testType) {
    default:
    case 'undefined':
      console.error('• No `testType` passed. Test type should be the passed - e.g. `testType=unitTest`.');
      break;
    case 'unitTest':
      {
        (0, _script.unitTest)({
          configuration,
          container: {
            imageName: namedArgs.imageName || process.env.imageName },

          nodeFlag: {
            debug: process.argv.includes('debug'),
            break: process.argv.includes('break') },

          testPath: namedArgs['path'],
          applicationPathOnHostMachine: process.env.applicationPathOnHostMachine });

      }
      break;}

}
